enterprise/users/src/main/java/io/crate/operation/user/TransportAlterUserAction.java

```diff
@@ -0,0 +1,114 @@
+import org.elasticsearch.common.Priority;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+
+public class TransportAlterUserAction extends TransportMasterNodeAction<AlterUserRequest, WriteUserResponse> {
+
+    @Inject
+    public TransportAlterUserAction(Settings settings,
+                                       TransportService transportService,
+                                       ClusterService clusterService,
+                                       ThreadPool threadPool,
+                                       ActionFilters actionFilters,
+                                       IndexNameExpressionResolver indexNameExpressionResolver) {
+        super(settings, "crate/sql/alter_user", transportService, clusterService, threadPool, actionFilters,
+            indexNameExpressionResolver, AlterUserRequest::new);
+    }
+
+    @Override
+    protected String executor() {
+        return ThreadPool.Names.GENERIC;
+    }
+
+    @Override
+    protected WriteUserResponse newResponse() {
+        return new WriteUserResponse(true);
+    }
+
+    @Override
+    protected void masterOperation(AlterUserRequest request, ClusterState state, ActionListener<WriteUserResponse> listener) {
+        clusterService.submitStateUpdateTask("alter_user [" + request.userName() + "]",
+            new AckedClusterStateUpdateTask<WriteUserResponse>(Priority.URGENT, request, listener) {
+
+                private boolean userExists = true;
+
+                @Override
+                public ClusterState execute(ClusterState currentState) throws Exception {
+                    MetaData currentMetaData = currentState.metaData();
+                    MetaData.Builder mdBuilder = MetaData.builder(currentMetaData);
+                    userExists = alterUser(
+                        mdBuilder,
+                        request.userName(),
+                        request.secureHash()
+                    );
+                    return ClusterState.builder(currentState).metaData(mdBuilder).build();
+                }
+
+                @Override
+                protected WriteUserResponse newResponse(boolean acknowledged) {
+                    return new WriteUserResponse(acknowledged, userExists);
+                }
+            });
+    }
+
+    @VisibleForTesting
+    static boolean alterUser(MetaData.Builder mdBuilder, String userName, @Nullable SecureHash secureHash) {
+        UsersMetaData oldMetaData = (UsersMetaData) mdBuilder.getCustom(UsersMetaData.TYPE);
+        if (oldMetaData == null || !oldMetaData.contains(userName)) {
+            return false;
+        }
+        // create a new instance of the metadata, to guarantee the cluster changed action.
+        UsersMetaData newMetaData = UsersMetaData.newInstance(oldMetaData);
+        newMetaData.add(userName, secureHash);
```

Comment:
maybe we should rename the `add()` method to `put` to express what's actually happening